### PR TITLE
Iss651

### PR DESCRIPTION
**Pull request by gorokhovnik**
1. Extending pipeline tuning testing with different models and metrics
2. Fix PipelineTuner.get_metric_value calculation for binary+rf+accuracy-like case (can't calculate metrics with exact labels while returning 1st class probability)
3. Simplified _greater_is_better code

### CODE REVIEW

File: test/unit/pipelines/test_pipeline_tuning.py [hunk 1]
```diff
@@ -270,21 +297,22 @@ def test_certain_node_tuning_classification_correct(data_fixture, request):
     train_data, test_data = train_test_data_setup(data=data)
 
     # Pipelines for classification task
-    pipeline_simple = get_simple_class_pipeline()
+    pipelines_simple = [get_simple_class_pipeline(ot) for ot in get_class_operation_types()]
     pipeline_complex = get_complex_class_pipeline()
 
-    for pipeline in [pipeline_simple, pipeline_complex]:
+    for pipeline in pipelines_simple + [pipeline_complex]:
```
Comment (nicl-nno) [inline test/unit/pipelines/test_pipeline_tuning.py:303]: Лучше всё-таки сделать это заранее.

File: test/unit/pipelines/test_pipeline_tuning.py [hunk 2]
```diff
@@ -243,21 +269,22 @@ def test_certain_node_tuning_regression_correct(data_fixture, request):
     train_data, test_data = train_test_data_setup(data=data)
 
     # Pipelines for regression task
-    pipeline_simple = get_simple_regr_pipeline()
+    pipelines_simple = [get_simple_regr_pipeline(ot) for ot in get_regr_operation_types()]
```
Comment (nicl-nno) [inline test/unit/pipelines/test_pipeline_tuning.py:272]: ot непонятно.

File: test/unit/pipelines/test_pipeline_tuning.py [hunk 3]
```diff
@@ -135,22 +163,21 @@ def test_pipeline_tuner_regression_correct(data_fixture, request):
...
-                                           task=train_data.task,
-                                           iterations=1,
-                                           search_space=search_space,
-                                           algo=tpe.suggest)
-            # Optimization will be performed on RMSE metric, so loss params are defined
-            tuned_pipeline = pipeline_tuner.tune_pipeline(input_data=train_data,
-                                                          loss_function=mse,
-                                                          loss_params={'squared': False})
-            assert pipeline_tuner.obtained_metric is not None
+            for loss in get_regr_losses():
```
Comment (nicl-nno) [inline test/unit/pipelines/test_pipeline_tuning.py:171]: А сколько по времени такой тест занимает по сравнению с предыдущей реализацией?

**Review by nicl-nno — CHANGES_REQUESTED** [review]
А где именно вот это:

Fix PipelineTuner.get_metric_value calculation for binary+rf+accuracy-like case (can't calculate metrics with exact labels while returning 1st class probability)

фиксится?

**Comment by gorokhovnik** [comment]
> А где именно вот это:
> 
> Fix PipelineTuner.get_metric_value calculation for binary+rf+accuracy-like case (can't calculate metrics with exact labels while returning 1st class probability)
> 
> фиксится?

Пока не фиксится, я только добавил тестов, чтобы воспроизвести баг. Позову на ревью, когда будет готово

File: test/unit/pipelines/test_pipeline_tuning.py (hunk 3 above)
Comment (gorokhovnik) [reply test/unit/pipelines/test_pipeline_tuning.py:171]: На локалке 8 секунд против 51. Возможно, на регрессии это имеет меньше смысла, потому что метки всегда возвращаются числом и в метриках всегда используется число. Но на классификации, кажется, essential

Comment (nicl-nno) [reply test/unit/pipelines/test_pipeline_tuning.py:171]: Тогда предлагаю оставить только одну регрессионную метрику и сделать поменьше регрессионных моделей - раз там маловероятно возникновение проблем.

File: test/unit/pipelines/test_pipeline_tuning.py (hunk 2 above)
Comment (gorokhovnik) [reply test/unit/pipelines/test_pipeline_tuning.py:272]: done

File: test/unit/pipelines/test_pipeline_tuning.py (hunk 1 above)
Comment (gorokhovnik) [reply test/unit/pipelines/test_pipeline_tuning.py:303]: done

File: test/unit/pipelines/test_pipeline_tuning.py (hunk 3 above)
Comment (gorokhovnik) [reply test/unit/pipelines/test_pipeline_tuning.py]: Done

**Comment by nicl-nno** [comment]
image

Кажется эти тесты действительно довольно однообразны. Може заменить на один тест с разными фикстурами?

File: fedot/core/pipelines/tuning/tuner_interface.py [hunk 4]
```diff
@@ -211,89 +211,53 @@ def _cross_validation(self, data, pipeline):
...
 
     :return: bool value is it good to maximize metric or not
     """
 
-    if isinstance(target[0], str):
-        # Target for classification contain string objects
-        le = LabelEncoder()
-        target = le.fit_transform(target)
+    optimal_target = np.array([[0], [1]])
+    not_optimal_target = np.array([[0], [0]])
```
Comment (nicl-nno) [inline fedot/core/pipelines/tuning/tuner_interface.py:247]: Не очень понятно, что тут имеется в виду и что происходит с ними дальше.

Comment (nicl-nno) [reply fedot/core/pipelines/tuning/tuner_interface.py:247]: Ок, я догадался. Но все равно стоит пояснить в комментариях и названиях переменных. Само понятие optimal_target не очень логичное.

Comment (gorokhovnik) [reply fedot/core/pipelines/tuning/tuner_interface.py]: Done

**Review by Dreamlone — COMMENTED** [review]
От себя отмечу, что изменения в самом коде ок, но тесты чет очень долго работают кмк
image

Можно их упростить? Можно например уменьшить количество перебираемых вариантов пайплайнов

**Comment by nicl-nno** [comment]
https://github.com/nccr-itmo/FEDOT/tree/iss-651-reduced-time

Вот сюда вкинул вариант упрощения.

**Review by nicl-nno — APPROVED** [review]
Время снизилось до 7-8 минут, норм.
